File: WithTable/app/src/main/java/com/pos_billingwala/Extra/LogSanitizer.java
package com.pos_billingwala.Extra;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Redacts secrets from crash/error payloads before queue/upload.
 * Masks credentials only — never replaces technical exception text with generics.
 */
public final class LogSanitizer {

    public static final int DEFAULT_MAX_BYTES = 20 * 1024;

    private static final Pattern JSON_SECRET = Pattern.compile(
            "(\"(?:password|mpin|otp|token|authorization|authToken|refresh_token|refreshToken|"
                    + "access_token|accessToken|aadhaar|aadhar|card_number|cardNumber|cvv|secret|"
                    + "api_key|apiKey|app_licence_key|licence_key|license_key|android_device_id|androidId)"
                    + "\"\\s*:\\s*\")([^\"]*)(\")",
            Pattern.CASE_INSENSITIVE
    );

    private static final Pattern FORM_SECRET = Pattern.compile(
            "((?:password|mpin|otp|token|authorization|authToken|refresh_token|refreshToken|"
                    + "access_token|accessToken|aadhaar|aadhar|card_number|cvv|secret|api_key|"
                    + "app_licence_key|licence_key|license_key|android_device_id|androidId)=)([^&\\s]*)",
            Pattern.CASE_INSENSITIVE
    );

    private static final Pattern BEARER = Pattern.compile(
            "(Bearer\\s+)(\\S+)",
            Pattern.CASE_INSENSITIVE
    );

    private LogSanitizer() {
    }

    public static String sanitize(String raw) {
        return sanitize(raw, DEFAULT_MAX_BYTES);
    }

    public static String sanitize(String raw, int maxBytes) {
        if (raw == null || raw.isEmpty()) {
            return "";
        }
        String sanitized = raw;
        sanitized = redactBearer(sanitized);
        sanitized = redactJson(sanitized);
        sanitized = redactForm(sanitized);
        if (maxBytes > 0 && sanitized.length() > maxBytes) {
            return sanitized.substring(0, maxBytes) + "…[truncated]";
        }
        return sanitized;
    }

    /** True if message is a forbidden generic placeholder (must not overwrite originals). */
    public static boolean isGenericMessage(String msg) {
        if (msg == null) {
            return false;
        }
        String n = msg.trim().toLowerCase();
        return n.equals("something went wrong")
                || n.equals("unknown error")
                || n.equals("operation failed")
                || n.equals("an error occurred")
                || n.equals("error occurred")
                || n.equals("failed")
                || n.equals("unknown");
    }

    private static String redactBearer(String input) {
        Matcher matcher = BEARER.matcher(input);
        StringBuffer out = new StringBuffer();
        while (matcher.find()) {
            matcher.appendReplacement(out, Matcher.quoteReplacement(matcher.group(1) + "********"));
        }
        matcher.appendTail(out);
        return out.toString();
    }

    private static String redactJson(String input) {
        Matcher matcher = JSON_SECRET.matcher(input);
        StringBuffer out = new StringBuffer();
        while (matcher.find()) {
            matcher.appendReplacement(out, Matcher.quoteReplacement(matcher.group(1) + "******" + matcher.group(3)));
        }
        matcher.appendTail(out);
        return out.toString();
    }

    private static String redactForm(String input) {
        Matcher matcher = FORM_SECRET.matcher(input);
        StringBuffer out = new StringBuffer();
        while (matcher.find()) {
            matcher.appendReplacement(out, Matcher.quoteReplacement(matcher.group(1) + "******"));
        }
        matcher.appendTail(out);
        return out.toString();
    }
}

File: API/log_sanitizer.php
<?php
/**
 * Redact secrets from crash/error log payloads before storage or Admin display.
 * Masks credentials only — does not replace technical error messages with generics.
 */

if (!function_exists('log_sanitize_text')) {
    /**
     * @param string|null $raw
     * @param int $maxBytes
     * @return string
     */
    function log_sanitize_text($raw, $maxBytes = 20480)
    {
        if ($raw === null || $raw === '') {
            return '';
        }
        $text = (string) $raw;

        // Bearer / Authorization headers
        $text = preg_replace('/(Bearer\s+)\S+/i', '$1********', $text);
        $text = preg_replace('/("?(?:authorization|auth[_-]?token|refresh[_-]?token|access[_-]?token)"?\s*[:=]\s*"?)([^"&\s,}]+)/i', '$1********', $text);

        // JSON-style secrets
        $jsonKeys = 'password|mpin|otp|token|authorization|refresh_token|access_token|aadhaar|aadhar|card_number|cardNumber|cvv|secret|api_key|apiKey|app_licence_key|licence_key|license_key';
        $text = preg_replace(
            '/("(?:' . $jsonKeys . ')"\s*:\s*")([^"]*)(")/i',
            '$1******$3',
            $text
        );

        // Form / query secrets
        $formKeys = 'password|mpin|otp|token|authorization|refresh_token|access_token|aadhaar|aadhar|card_number|cvv|secret|api_key|app_licence_key|licence_key|license_key|authToken';
        $text = preg_replace(
            '/((?:' . $formKeys . ')=)([^&\s]*)/i',
            '$1******',
            $text
        );

        if ($maxBytes > 0 && strlen($text) > $maxBytes) {
            $text = substr($text, 0, $maxBytes) . '…[truncated]';
        }
        return $text;
    }
}

if (!function_exists('log_post_str')) {
    /**
     * @param string $key
     * @param int $maxBytes
     * @return string
     */
    function log_post_str($key, $maxBytes = 20480)
    {
        $val = isset($_POST[$key]) ? $_POST[$key] : '';
        return log_sanitize_text($val, $maxBytes);
    }
}

if (!function_exists('log_is_generic_message')) {
    /**
     * Detect forbidden generic replacements that must never overwrite originals.
     * @param string $msg
     * @return bool
     */
    function log_is_generic_message($msg)
    {
        $normalized = strtolower(trim((string) $msg));
        if ($normalized === '') {
            return false;
        }
        $generics = array(
            'something went wrong',
            'unknown error',
            'operation failed',
            'an error occurred',
            'error occurred',
            'failed',
            'unknown',
        );
        return in_array($normalized, $generics, true);
    }
}
